### PR TITLE
Make Camo Tarps Great Again

### PR DESCRIPTION
**Pull request by Surrealaser**
## About The Pull Request

Per the changelog.

## Why It's Good For The Game

Makes camo tarps great again for ambushes and guerrilla tactics rather than being nearly worthless and far too situational.

## Changelog
:cl:
tweak: Xenos no longer auto-open camo tarps by moving over them.
tweak: Acid spray and gas, bullets (including spits) and phosphorous gas now all force camo tarp occupants out and affect them. Projectiles must specifically target the tarp.
tweak: Camo tarps now stun their former occupant for 0.5 upon leaving.
refactor: Minor refactors of camo tarps.
/:cl:

### CODE REVIEW

**Comment by SplinterGP** [comment]
Hmm, i like it, just the problems i see is, bullets opening it will cause big FF and xenos not opening on passing over is cool but i think it only opening on spit gas cloud etc, a situational downgrade since bullets or anything will openning it meanwhile seeing by that none of the combat castes except from sentinels evolves would be able to deal with tarps.

**Comment by Surrealaser** [comment]
> Hmm, i like it, just the problems i see is, bullets opening it will cause big FF and xenos not opening on passing over is cool but i think it only opening on spit gas cloud etc, a situational downgrade since bullets or anything will openning it meanwhile seeing by that none of the combat castes except from sentinels evolves would be able to deal with tarps.

You have to actually shoot the bag to hit it with a bullet.

**Comment by SplinterGP** [comment]
Ah well shoot the bag itself got.it

**Comment by AbsoluteJerry** [comment]
I FEEL YOUNG AGAIN. THE TARP IS BACK IN THE GAME BOYYYS!!

File: code/game/objects/effects/aliens.dm [hunk 1]
```diff
@@ -97,6 +97,12 @@
 			continue
 		Crossed(M)
 
+	for(var/atom/A in loc) //Infrastructure for other interactions
+		A.acidspray_act()
```
Comment (spookydonut) [inline code/game/objects/effects/aliens.dm:101]:
```suggestion
		SEND_SIGNAL(A, COMSIG_ATOM_ACID_SPRAYED, src)
```

File: code/game/objects/items/marine_gear.dm [hunk 2]
```diff
@@ -18,11 +18,28 @@
...
 	unfoldedbag_path = /obj/structure/closet/bodybag/tarp
+	var/serial_number //Randomized serial number used to stop point macros and such.
+
+/obj/item/bodybag/tarp/Initialize(mapload, unfoldedbag)
+	. = ..()
+	sleep(1)
+	if(!serial_number)
+		serial_number = "SN-[rand(1000,100000)]" //Set the serial number
+		name = "\improper [serial_number] [name]"
+
```
Comment (spookydonut) [inline code/game/objects/items/marine_gear.dm:29]:
```suggestion
	RegisterSignal(src, COMSIG_ATOM_ACID_SPRAYED, .proc/acidspray_act)
```

File: code/game/objects/items/marine_gear.dm [hunk 3]
```diff
@@ -87,74 +105,63 @@
...
-		M.bullet_act(proj) //tarp isn't bullet proof; concealment, not cover; pass it on to the occupant.
+/obj/structure/closet/bodybag/bullet_act(obj/projectile/proj)
+	if(!opened && bodybag_occupant)
+		bodybag_occupant.bullet_act(proj) //tarp isn't bullet proof; concealment, not cover; pass it on to the occupant.
+		to_chat(bodybag_occupant, "<span class='danger'>The projectile disrupts the camoflague [sanitize(src.name)]!</span>")
+		open() //Yeah, this is a bit harder to justify, but better than snowflaking a loss of alpha.
 
-/obj/structure/closet/bodybag/tarp/Crossed(mob/living/L)
+/obj/structure/closet/bodybag/acidspray_act()
 	. = ..()
```
Comment (spookydonut) [inline code/game/objects/items/marine_gear.dm:146]:
```suggestion
/obj/structure/closet/bodybag/proc/acidspray_act()
```

File: code/game/objects/items/marine_gear.dm [hunk 4]
```diff
@@ -87,74 +105,63 @@
...
+	if(!.)
+		return
 
-/obj/effect/tarp_trigger/Crossed(atom/A)
-	. = ..()
-	if(!linked_tarp) //something went very wrong
-		qdel(src)
-	else if(isxeno(A))
-		addtimer(CALLBACK(linked_tarp, /obj/structure/closet/bodybag/tarp.proc/trigger_open), 5 SECONDS)
+	if(CHECK_BITFIELD(S.smoke_traits, SMOKE_XENO_ACID || SMOKE_BLISTERING ) && !opened && bodybag_occupant)
```
Comment (spookydonut) [inline code/game/objects/items/marine_gear.dm:161]:
```suggestion
	if((S.smoke_traits & SMOKE_XENO_ACID|SMOKE_BLISTERING) && !opened && bodybag_occupant)
```

**Review by spookydonut — CHANGES_REQUESTED** [review]

File: code/game/objects/effects/aliens.dm [hunk 5]
```diff
@@ -97,6 +97,12 @@
 			continue
 		Crossed(M)
 
+	for(var/atom/A in loc) //Infrastructure for other interactions
+		SEND_SIGNAL(A, COMSIG_ATOM_ACID_SPRAYED, src)
+
+/atom/proc/acidspray_act()
+	return
```
Comment (spookydonut) [inline code/game/objects/effects/aliens.dm:104]:
```suggestion
```

File: code/game/objects/items/marine_gear.dm [hunk 6]
```diff
@@ -18,11 +18,28 @@
...
 	unfoldedbag_path = /obj/structure/closet/bodybag/tarp
+	var/serial_number //Randomized serial number used to stop point macros and such.
+
+/obj/item/bodybag/tarp/Initialize(mapload, unfoldedbag)
+	. = ..()
+	sleep(1)
+	if(!serial_number)
+		serial_number = "SN-[rand(1000,100000)]" //Set the serial number
+		name = "\improper [serial_number] [name]"
+	RegisterSignal(src, COMSIG_ATOM_ACID_SPRAYED, .proc/acidspray_act)
```
Comment (spookydonut) [inline code/game/objects/items/marine_gear.dm:29]:
```suggestion
```

File: code/game/objects/items/marine_gear.dm [hunk 7]
```diff
@@ -40,9 +57,11 @@
 	open_sound = 'sound/effects/vegetation_walk_1.ogg'
 	close_sound = 'sound/effects/vegetation_walk_2.ogg'
 	foldedbag_path = /obj/item/bodybag/tarp
-	closet_stun_delay = 0
-	var/list/tarp_triggers = list()
+	closet_stun_delay = 0.5 SECONDS //Short delay to prevent ambushes from being too degenerate.
+	var/serial_number //Randomized serial number used to stop point macros and such.
 
+/obj/structure/closet/bodybag/tarp/Initialize(mapload, foldedbag)
+	. = ..()
```
Comment (spookydonut) [inline code/game/objects/items/marine_gear.dm:64]:
```suggestion
/obj/structure/closet/bodybag/Initialize(mapload, foldedbag)
	. = ..()
	RegisterSignal(src, COMSIG_ATOM_ACID_SPRAYED, .proc/acidspray_act)
```

**Review by spookydonut — CHANGES_REQUESTED** [review]

File: code/game/objects/items/bodybag.dm [hunk 8]
```diff
@@ -210,6 +209,56 @@
 	if(src != proj.original_target) //You miss unless you click directly on the bodybag
 		return FALSE
 
+	if(!opened && bodybag_occupant)
+		bodybag_occupant.bullet_act(proj) //tarp isn't bullet proof; concealment, not cover; pass it on to the occupant.
+		to_chat(bodybag_occupant, "<span class='danger'>You jolt out of [sanitize(src.name)] upon being hit!</span>")
+		open()
+
+
+/obj/structure/closet/bodybag/fire_act(exposed_temperature, exposed_volume)
```
Comment (spookydonut) [inline code/game/objects/items/bodybag.dm:218]: fire_act() is never called its legacy atmos fire stuff.

File: code/game/objects/items/bodybag.dm [hunk 9]
```diff
@@ -210,6 +209,56 @@
...
+			visible_message("<span class='danger'>\The shockwave blows [sanitize(src.name)] apart!</span>")
+			qdel(src) //blown apart
+
+/obj/structure/closet/bodybag/proc/acidspray_act()
+	if(!opened && bodybag_occupant)
+		var/obj/effect/xenomorph/spray/S = locate() in range(0, src) //get the acid Hans
+		if(!S) //Sanity
+			return
+
+		S.Crossed(bodybag_occupant) //tarp isn't acid proof; pass it on to the occupant
```
Comment (spookydonut) [inline code/game/objects/items/bodybag.dm:247]: you shouldnt do this, calling Crossed directly

**Review by spookydonut — CHANGES_REQUESTED** [review]

**Comment by Surrealaser** [comment]
For those who are curious why progress isn't being made on this, currently waiting on https://github.com/tgstation/TerraGov-Marine-Corps/pull/5174 to merge.

**Comment by Hughgent** [comment]
Please 'return to full PR' status when you are ready for another review.

**Review by Hughgent — APPROVED** [review]
Nothing immediately strikes me as wrong.

@spookydonut please set the TMC tag if you also approve.

File: code/game/objects/items/bodybag.dm [hunk 10]
```diff
@@ -320,3 +366,116 @@
...
+	icon_closed = "jungletarp_closed"
+	icon_opened = "jungletarp_open"
+	open_sound = 'sound/effects/vegetation_walk_1.ogg'
+	close_sound = 'sound/effects/vegetation_walk_2.ogg'
+	foldedbag_path = /obj/item/bodybag/tarp
+	closet_stun_delay = 0.5 SECONDS //Short delay to prevent ambushes from being too degenerate.
+	var/serial_number //Randomized serial number used to stop point macros and such.
+
+/obj/structure/closet/bodybag/tarp/Initialize(mapload, foldedbag)
+	. = ..()
```
Comment (spookydonut) [inline code/game/objects/items/bodybag.dm:423]:
```suggestion
```

File: code/game/objects/items/bodybag.dm [hunk 11]
```diff
@@ -210,6 +211,51 @@
...
+	if(!opened && bodybag_occupant)
+		to_chat(bodybag_occupant, "<span class='danger'>The shockwave blows [sanitize(src.name)] open!</span>")
+		open()
+		bodybag_occupant.ex_act(severity)
+	switch(severity)
+		if(EXPLODE_DEVASTATE)
+			visible_message("<span class='danger'>\The shockwave blows [sanitize(src.name)] apart!</span>")
+			qdel(src) //blown apart
+
+/obj/structure/closet/bodybag/proc/acidspray_act()
```
Comment (spookydonut) [inline code/game/objects/items/bodybag.dm:235]:
```suggestion
/obj/structure/closet/bodybag/proc/acidspray_act()
	SIGNAL_HANDLER
```

**Review by spookydonut — CHANGES_REQUESTED** [review]

File: code/game/objects/items/bodybag.dm [hunk 12]
```diff
@@ -210,6 +211,52 @@
...
+	SIGNAL_HANDLER
+	if(!opened && bodybag_occupant)
+		var/obj/effect/xenomorph/spray/S = locate() in range(0, src) //get the acid Hans
+		if(!S) //Sanity
+			return
+
+		if(ishuman(bodybag_occupant))
+			var/mob/living/carbon/human/H = bodybag_occupant
+			H.acid_spray_crossed(S.slow_amt) //tarp isn't acid proof; pass it on to the occupant
```
Comment (spookydonut) [inline code/game/objects/items/bodybag.dm:244]:
```suggestion
			INVOKE_ASYNC(H, /mob/living/carbon/human.proc/acid_spray_crossed, S.slow_amt) //tarp isn't acid proof; pass it on to the occupant
```

**Review by spookydonut — APPROVED** [review]

**Comment by MiniMeatwad** [comment]
Add back the opening whenever a xenomorph steps over it. It's situational and requires you to place it somewhere xenomorphs wont walk over for a reason. After all it is invisibility with no drawbacks.

**Pull request merged**